src/ray/gcs/store_client/in_memory_store_client.h

```diff
@@ -73,25 +74,26 @@ class InMemoryStoreClient : public StoreClient {
  private:
   struct InMemoryTable {
     /// Mutex to protect the records_ field and the index_keys_ field.
-    absl::Mutex mutex_;
+    mutable absl::Mutex mutex_;
     // Mapping from key to data.
+    // TODO(dayshah): benchmark reader/writer locks against boost::concurrent_flat_map
     absl::flat_hash_map<std::string, std::string> records_ ABSL_GUARDED_BY(mutex_);
   };
 
-  std::shared_ptr<InMemoryStoreClient::InMemoryTable> GetOrCreateTable(
-      const std::string &table_name) ABSL_LOCKS_EXCLUDED(mutex_);
+  InMemoryTable &CreateOrGetMutableTable(const std::string &table_name);
+
+  std::optional<const InMemoryTable *> GetTable(const std::string &table_name);
 
   /// Mutex to protect the tables_ field.
   absl::Mutex mutex_;
-  absl::flat_hash_map<std::string, std::shared_ptr<InMemoryTable>> tables_
-      ABSL_GUARDED_BY(mutex_);
+  absl::node_hash_map<std::string, InMemoryTable> tables_ ABSL_GUARDED_BY(mutex_);
```

Comment:
can you add a line of comment explaining why we use node_hash_map ?

Reply:
added